lib/bolt/executor.rb

```diff
@@ -373,83 +382,156 @@ def run_plan(scope, plan, params)
+
+      # Compress the current scopes into a single vars hash to add to the new scope
+      scope.to_hash.each_pair { |k, v| local[k] = v }
+      newscope.push_ephemerals([local])
+
       fiber = Fiber.new do
-        # Create the new scope
-        newscope = Puppet::Parser::Scope.new(scope.compiler)
-        local = Puppet::Parser::Scope::LocalScope.new
-
-        # Compress the current scopes into a single vars hash to add to the new scope
-        current_scope = scope.effective_symtable(true)
-        until current_scope.nil?
-          current_scope.instance_variable_get(:@symbols)&.each_pair { |k, v| local[k] = v }
-          current_scope = current_scope.parent
+        result = catch(:return) do
+          args = { block.parameters[0][1].to_s => object }
+          block.closure.call_by_name_with_scope(newscope, args, true)
         end
-        newscope.push_ephemerals([local])
 
-        begin
-          result = catch(:return) do
-            args = { block.parameters[0][1].to_s => object }
-            block.closure.call_by_name_with_scope(newscope, args, true)
-          end
+        # If we got a return from the block, get it's value
+        # Otherwise the result is the last line from the block
+        result = result.value if result.is_a?(Puppet::Pops::Evaluator::Return)
 
-          # If we got a return from the block, get it's value
-          # Otherwise the result is the last line from the block
-          result = result.value if result.is_a?(Puppet::Pops::Evaluator::Return)
+        # Validate the result is a PlanResult
+        unless Puppet::Pops::Types::TypeParser.singleton.parse('Boltlib::PlanResult').instance?(result)
+          raise Bolt::InvalidParallelResult.new(result.to_s, *Puppet::Pops::PuppetStack.top_of_stack)
+        end
 
-          # Validate the result is a PlanResult
-          unless Puppet::Pops::Types::TypeParser.singleton.parse('Boltlib::PlanResult').instance?(result)
-            raise Bolt::InvalidParallelResult.new(result.to_s, *Puppet::Pops::PuppetStack.top_of_stack)
-          end
+        result
+      rescue Puppet::PreformattedError => e
+        if e.cause.is_a?(Bolt::Error)
+          e.cause
+        else
+          raise e
+        end
+      end
 
-          result
-        rescue Puppet::PreformattedError => e
-          if e.cause.is_a?(Bolt::Error)
-            e.cause
+      fiber = Bolt::Future.new(fiber, name)
+      @logger.trace("Created future #{name}") if in_parallel
+      plan_fibers.fibers << fiber
+      fiber
+    end
+
+    # Check on all the Fibers and see if any can be resumed.
+    def round_robin(rrfibers = plan_fibers.all)
+      r = nil
+
+      rrfibers.each do |fiber|
+        # If the Fiber is still running and can be resumed, then resume it and
+        # save the output to check if the task has finished or not.
+        @logger.trace("Checking future '#{fiber.name}'") if in_parallel
+        if fiber.alive?
```

Comment:
In general fibers should only become not alive during a resume call. I think this if-else confuses things and forces us into an "extra" loop when things finish. It seems like we should expect fibers to be alive, call `resume`, and then check to see if they completed and do the cleanup each loop rather than doing one then the other the next loop.